### PR TITLE
s390x: Migrate branches and traps to ISLE

### PR DESCRIPTION
**Pull request by uweigand**
In order to migrate branches to ISLE, we define a second entry
point `lower_branch` which gets the list of branch targets as
additional argument.

This requires a small change to `lower_common`: the `isle_lower`
callback argument is changed from a function pointer to a closure.
This allows passing the extra argument via a closure.

Traps make use of the recently added facility to emit safepoints
from ISLE, but are otherwise straightforward.

@cfallin @fitzgen

### CODE REVIEW

File: cranelift/codegen/src/isa/s390x/inst.isle
Comment (fitzgen) [inline cranelift/codegen/src/isa/s390x/inst.isle:2201]: I think this should go in the shared prelude, next to `value_regs_none`.

File: cranelift/codegen/src/isa/s390x/lower/isle.rs
Comment (fitzgen) [inline cranelift/codegen/src/isa/s390x/lower/isle.rs:510]: And then I would expect this to go into `wasmtime/cranelift/codegen/src/machinst/isle.rs`

File: cranelift/codegen/src/isa/s390x/inst.isle
Comment (fitzgen) [inline cranelift/codegen/src/isa/s390x/inst.isle:1300]: And this should be in the prelude as well.

**Review by fitzgen — COMMENTED** [review]
Looks great modulo a couple nitpicks below, thanks!

File: cranelift/codegen/src/isa/s390x/inst.isle
Comment (fitzgen) [reply cranelift/codegen/src/isa/s390x/inst.isle:1300]: I see that `emit` actually isn't in the prelude, but we do *use* it from the prelude. So we should probably move `emit` to the prelude as well, since in practice it has to be the same declaration across all backends anyways. Happy to have that as part of this updated PR or we can do it in another PR.

Comment (uweigand) [reply cranelift/codegen/src/isa/s390x/inst.isle:1300]: Yes, that's why I didn't put it in the prelude ...

I agree it makes sense to move emit / emit_safepoint there, but then we should also move the implementations to machinst/isle.re, right?

I'd be happy to do that, but I think it would be better to do it in another PR, as it will touch all back-ends.

Comment (uweigand) [reply cranelift/codegen/src/isa/s390x/inst.isle:1300]: Ah, I see.   But even if we don't move the implementaton, once the `decl` is in the prelude as `extern constructor`, *every* back-end *must* implement it or it won't compile any more.   So we'll still need to touch all back-ends.  (In fact, with the x86 `mov_mitosis` implementation, I'm not sure exactly how the safepoint variant would work.  I guess only the last instruction in the list should be the safepoint?)

Comment (fitzgen) [reply cranelift/codegen/src/isa/s390x/inst.isle:1300]: > once the `decl` is in the prelude as `extern constructor`, _every_ back-end _must_ implement it or it won't compile any more.

Correct. (Also, it is already the case that every backend must implement it or the ISLE won't compile because we are calling `emit` and assuming it is declared from inside the prelude already, it just so happens that we have N identical declarations for each backend instead of a single shared declaration in the prelude.)

> (In fact, with the x86 `mov_mitosis` implementation, I'm not sure exactly how the safepoint variant would work. I guess only the last instruction in the list should be the safepoint?)

Yeah, I will have to think a bit about how to do this best. I think making only the last instruction a safepoint would work, but we might want to have some debug asserts in there too. Planning on tackling this after I'm done with my newtype wrappers PR by which time this PR will have merged.

Comment (cfallin) [reply cranelift/codegen/src/isa/s390x/inst.isle:1300]: > I think making only the last instruction a safepoint would work,

Yeah, I think that's the correct solution -- the expanded list should always have the form of zero or more moves, followed by the original instruction; and the safepoint bit applies to the latter.

Comment (fitzgen) [reply cranelift/codegen/src/isa/s390x/inst.isle:1300]: There are some instructions that always generate their results into a particular register, and so when our SSA-ified version of that instruction wants the result in a temp, we have to generate a move that follows the "real" instruction, so we can have trailing moves as well after move mitosis. This shouldn't be the case for safepoints I don't think (except maybe ABI/calling convention stuff where results are always in `rax`?) and we should be able to just debug assert this.

**Comment by uweigand** [comment]
Can this be merged now or are you waiting on any action from my side?   I have other patches I'd like to submit and this is causing merge conflicts ...

As to the question of moving `emit_safepoint` to the common prelude, my preference would still be for other architecture maintainers to implement this for their platform as fits them best, and once it is done for all, we can still merge any definitions that end up the same back into the common prelude ...

**Comment by uweigand** [comment]
> > Can this be merged now or are you waiting on any action from my side?
> 
> I think just the nitpicks that we didn't already discuss and agree could wait for follow ups:
> 
>     * moving `emit_safepoint` into the prelude, and
> 
>     * moving its external rust implementation into `wasmtime/cranelift/codegen/src/machinst/isle.rs`.

But that was just what we discussed: As I understand, x64 will need a _different_ `emit_safepoint` implementation due the move mitosis stuff, so it's not clear this can be shared.

**Comment by fitzgen** [comment]
Ah right. Then just the `decl` and `extern` bits need to move to `prelude.isle` and the implementations can be pre-backend. We can stub out the other backends with `unimplemented!()` for now.

**Comment by uweigand** [comment]
> Ah right. Then just the `decl` and `extern` bits need to move to `prelude.isle` and the implementations can be pre-backend. We can stub out the other backends with `unimplemented!()` for now.

I'm happy to do that, but I'd really prefer this to be a separate PR.  That will still touch all backends and their generated files, and I'd prefer not to entangle this with this back-end change, which is already quite large as-is ...

**Review by fitzgen — APPROVED** [review]
Okay, we can do that in follow ups then.

**Comment by uweigand** [comment]
> Okay, we can do that in follow ups then.

Thanks!  The follow up is now here:  https://github.com/bytecodealliance/wasmtime/pull/3745

**Pull request merged**
